<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>cellula v. 0.0.900 - Trajectory inference</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">cellula v. 0.0.900</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-menu" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Menu</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-menu">    
        <li>
    <a class="dropdown-item" href="./index.html">
 <span class="dropdown-text">Home</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./install.html">
 <span class="dropdown-text">Installing cellula</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./usage.html">
 <span class="dropdown-text">Usage</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./plotting.html">
 <span class="dropdown-text">Plotting</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./parallelization.html">
 <span class="dropdown-text">Parallelization</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./clustering.html">
 <span class="dropdown-text">Clustering</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./identities.html">
 <span class="dropdown-text">Assigning identities</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./trajectories.html">
 <span class="dropdown-text">Trajectory inference</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./downsampling.html">
 <span class="dropdown-text">Downsampling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./references.html">
 <span class="dropdown-text">References</span></a>
  </li>  
    </ul>
  </li>
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#metacells" id="toc-metacells" class="nav-link active" data-scroll-target="#metacells">Metacells</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Trajectory inference</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>At the time of writing <code>cellula</code> only implements a wrapper around the <code>slingshot</code> method<a href="references.html#18">[18]</a><a id="18"></a> and the <code>monocle3</code> method <a href="references.html#19">[19]</a><a id="19"></a> for pseudotemporal trajectory inference, and the <code>testPseudotime()</code> method from <code>TSCAN</code><a href="references.html#20">[20]</a><a id="20"></a> for differential expression along a trajectory. More trajectory inference and differential expression methods will be implemented in the future.</p>
<p>The <code>monocle3</code> method has been originally developed to work on 2D embeddings such as UMAP, but given the distortion introduced by these embeddings this function allows users to input any embedding (such as PCA). Moreover, the <code>monocle3</code> method originally allows partitions to be specified. In the current implementation partitions have been disabled.</p>
<p>The <code>findTrajectories()</code> function takes a <code>SingleCellExperiment</code> object as input, and requires the user to specify the method (one of <code>"slingshot"</code> or <code>"monocle"</code>), the cluster label which will be used as an input to the MST creation in <code>slingshot</code>, or to identify the starting node in <code>monocle3</code>.</p>
<p>Other important parameters are:</p>
<ul>
<li><p><code>dr</code> - the reduced dimensional embedding in which trajectories will be estimated (default is “PCA”)</p></li>
<li><p><code>start</code> - the starting cluster for trajectory estimation, defaulting to “auto” for an entropy-based method</p></li>
<li><p><code>omega</code> - <code>slingshot</code> only. Whether or not to use a synthetic cluster to estimate disjointed trajectories, as detailed in <code>?slingshot::getLineages</code></p></li>
<li><p><code>Monocle_lg_control</code> - <code>monocle</code> only. A list of control parameters for the <code>learn_graph()</code> function from <code>monocle3</code>.</p></li>
<li><p><code>invert</code> - <code>monocle</code> only. Whether or not to invert the direction of the pseudotime vector.</p></li>
<li><p><code>dr_embed</code> - an additional 2D dimensional reduction slot in which to embed the trajectories (<code>slingshot</code> curves or <code>monocle</code> principal graph). Yields slightly different results according to the method.</p></li>
<li><p><code>doDE</code> - whether or not to perform differential expression on every trajectory.</p></li>
</ul>
<p>The output is the same object used in the input, with some additional fields.</p>
<ul>
<li><p>the <code>colData</code> slot will contain a column containing the <code>slingPseudotime_N</code> pseudotime values for each lineage, where N is the number of the lineage (<code>slingshot</code>) or just a column named <code>monoclePseudotime</code> with pseudotime values for each cell (<code>monocle</code>).</p></li>
<li><p>the <code>metadata</code> slot will several new elements depending on the method.</p>
<p>For <code>slingshot</code>:</p>
<ul>
<li><code>Slingshot_lineages</code>: the output of <code>slingLineages()</code>, i.e.&nbsp;the result of the MST construction</li>
<li><code>Slingshot_embedded_curves</code>: the output of <code>embedCurves()</code>, i.e. the projection of low-dimensional principal curves onto a 2D embedding such as UMAP. Optional and only created if the parameter <code>dr_embed</code> is an available dimensionality reduction slot in the object.</li>
<li><code>Slingshot_MST</code>: the MST used to determine lineages by <code>slingshot</code>. Optional and only added if the parameter <code>add_metadata</code> is set to TRUE.</li>
<li><code>Slingshot_curves</code>: the principal curves constructed by <code>slingshot</code>. Optional and only added if the parameter <code>add_metadata</code> is set to TRUE.</li>
<li><code>Slingshot_weights</code>: the weights assigned per cell within each lineage by <code>slingshot</code>. Optional and only added if the parameter <code>add_metadata</code> is set to TRUE.</li>
<li><code>Slingshot_params</code>: the params used in the <code>slingshot</code> call. Optional and only added if the parameter <code>add_metadata</code> is set to TRUE.</li>
<li><code>pseudotime_DE</code>: the results of the differential expression test. This is a list where each lineage is tested separately. Optional and only created if <code>doDE</code> is set to TRUE.</li>
</ul>
<p>For <code>monocle</code>:</p>
<ul>
<li><code>Monocle_embedded_curves</code>: the projection of the nodes of the principal graph to their nearest neighbor in 2D. Optional and only created if the parameter <code>dr_embed</code> is an available dimensionality reduction slot in the object.</li>
<li><code>Monocle_principal_graph</code>: the <code>igraph</code> object with the principal graph.</li>
<li><code>Monocle_principal_graph_aux</code>: the rest of the <code>CellDataSet</code> object with additional information on the principal graph.</li>
</ul></li>
</ul>
<p>In this example we download a dataset that captures early haematopoietic differentiation from Nestorowa et al.&nbsp;<a href="#21">[21]</a><a id="21"></a> and we do a quick processing and clustering.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>sce2 <span class="ot">=</span> scRNAseq<span class="sc">::</span><span class="fu">NestorowaHSCData</span>()</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>sce2 <span class="ot">=</span> <span class="fu">cellula</span>(sce2, <span class="at">name =</span> <span class="st">"nestorowa"</span>, <span class="at">do_qc =</span> <span class="cn">FALSE</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>sce2 <span class="ot">=</span> <span class="fu">makeGraphsAndClusters</span>(sce2, <span class="at">space =</span> <span class="fu">reducedDim</span>(sce2, <span class="st">"PCA"</span>)[,<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>])</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_UMAP</span>(sce2, <span class="at">umap_slot =</span> <span class="st">"UMAP"</span>, <span class="at">color_by =</span> <span class="st">"SNN_0.64"</span>, <span class="at">label_by =</span> <span class="st">"SNN_0.64"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="https://github-production-user-asset-6210df.s3.amazonaws.com/21171362/242682308-dc8b13c4-f4ea-4dc1-93ff-e169ae1b64de.png" width="400"></p>
<p>Then we take an arbitrary starting point (cluster 7) and calculate trajectories with both <code>slingshot</code> and <code>monocle</code> using the same space (PCA, first 20 components).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>sce2 <span class="ot">=</span> <span class="fu">findTrajectories</span>(sce2, <span class="at">dr =</span> <span class="st">"PCA"</span>, </span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">method =</span> <span class="st">"monocle"</span>,</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">ndims =</span> <span class="dv">20</span>, <span class="at">clusters =</span> <span class="st">"SNN_0.64"</span>, </span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">dr_embed =</span> <span class="st">"UMAP"</span>, <span class="at">start =</span> <span class="dv">7</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>                        </span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>sce2 <span class="ot">=</span> <span class="fu">findTrajectories</span>(sce2, <span class="at">dr =</span> <span class="st">"PCA"</span>, </span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>                        <span class="at">method =</span> <span class="st">"slingshot"</span>,</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>                        <span class="at">ndims =</span> <span class="dv">20</span>, <span class="at">clusters =</span> <span class="st">"SNN_0.64"</span>, </span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>                        <span class="at">dr_embed =</span> <span class="st">"UMAP"</span>, <span class="at">start =</span> <span class="dv">7</span>)           </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can visualize results on the UMAP since we specified a <code>dr_embed</code> parameter. Notice how <code>slingshot</code> will identify different pseudotimes (one per lineage) while <code>monocle3</code> will identify a single pseudotime. Moreover, the <code>plot_UMAP()</code> function can include a <code>trajectory</code> argument which points to the name of the element in the <code>metadata</code> slot containing segments to draw trajectories.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a> <span class="fu">plot_UMAP</span>(sce2, <span class="at">umap_slot =</span> <span class="st">"UMAP"</span>, <span class="at">color_by =</span> <span class="st">"slingPseudotime_1"</span>, </span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>           <span class="at">label_by =</span> <span class="st">"SNN_0.64"</span>, <span class="at">trajectory =</span> <span class="st">"Slingshot_embedded_curves"</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a> <span class="fu">plot_UMAP</span>(sce2, <span class="at">umap_slot =</span> <span class="st">"UMAP"</span>, <span class="at">color_by =</span> <span class="st">"monoclePseudotime"</span>, </span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">label_by =</span> <span class="st">"SNN_0.64"</span>, <span class="at">trajectory =</span> <span class="st">"Monocle_embedded_curves"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="https://github-production-user-asset-6210df.s3.amazonaws.com/21171362/242682268-8855a6ca-c106-4f23-b2e6-c54fdcedac64.png" width="400"></p>
<p><img src="https://github-production-user-asset-6210df.s3.amazonaws.com/21171362/242682289-8b43ed43-ab82-4d03-901a-d7109d3a3d52.png" width="400"></p>
<p>Since the 2D embedding of <code>monocle3</code> PCA-derived trajectories may be hard to understand, given the distortions introduced by UMAP, <code>cellula</code> includes an additional 2D embedding method, <code>dr_embed = "FR"</code>, inspired båy the PAGA embedding initialization technique <a href="references.html#22">[22]</a><a id="22"></a>.</p>
<p>Briefly, once the principal graph has been calculated, it is laid out in 2D using the Fruchterman-Reingold algorithm. Then each cell is randomly plotted around their closest vertex in the graph, and reordered according to pseudotime value. This semi-random layout is used as an initialization for UMAP, which will optimize the point positions. The resulting layout is more visually pleasing and reflects more accurately the positions of cells with respect to the trajectory (although not necessarily to each other).</p>
<p>This FR-initialized UMAP is stored in a <code>reducedDim</code> slot named <code>UMAP_FR</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>sce2 <span class="ot">=</span> <span class="fu">findTrajectories</span>(sce2, <span class="at">clusters =</span> <span class="st">"SNN_0.64"</span>, <span class="at">method =</span> <span class="st">"monocle"</span>,</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">dr =</span> <span class="st">"PCA"</span>, <span class="at">ndims =</span> <span class="dv">20</span>, <span class="at">start =</span> <span class="st">"7"</span>, <span class="at">dr_embed =</span> <span class="st">"FR"</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_UMAP</span>(sce2, <span class="at">umap_slot =</span> <span class="st">"UMAP_FR"</span>, <span class="at">color_by =</span> <span class="st">"monoclePseudotime"</span>, </span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>           <span class="at">label_by =</span> <span class="st">"SNN_0.64"</span>, <span class="at">trajectories =</span> <span class="st">"Monocle_embedded_curves"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="https://github-production-user-asset-6210df.s3.amazonaws.com/21171362/243069896-c1c933d0-8f23-4d73-b25c-5374eed73cb2.png" width="400"></p>
<p>It should be noted that this layout is, in a way, optimized for trajectories rather than global cell-cell similarity and as always should only be treated as a visualization tool.</p>
<section id="metacells" class="level2">
<h2 class="anchored" data-anchor-id="metacells">Metacells</h2>
<p>In order to speed up calculations and overcome sparsity, cells can be aggregated into metacells using k-means clustering with a high <em>k</em>.</p>
<p>Clustering is carried out in the reduced dimensional space of choice selected through the <code>space</code> argument, and it is carried out in each level of <code>group</code> separately.</p>
<p>Rather than selecting the number of clusters <em>k</em>, the function takes an average number of cells per cluster <em>w</em> which is used to determine <em>k</em> (default is w = 10 cells per cluster). Read counts are aggregated by gene across all cells within each cluster, resulting in metacells. These can be used as input to <code>findTrajectories()</code> or other operations such as clustering, downsampling, signature scoring, etc.</p>
<p>In this example we create metacells aggregating (on average) 10 cells, within each cluster from the “SNN_0.5” clustering results:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>sce_meta <span class="ot">&lt;-</span> <span class="fu">makeMetacells</span>(sce, <span class="at">group =</span> <span class="st">"SNN_0.5"</span>, <span class="at">space =</span> <span class="st">"PCA_Harmony"</span>, <span class="at">w =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>