<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-0.9.649">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Single Cell Analysis with R - 3&nbsp; Data input</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./qc.html" rel="next">
<link href="./pipelinechoice.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data input</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Single Cell Analysis with R</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Summary</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pipelinechoice.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Choice of pipeline</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./datainput.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data input</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./qc.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Quality Control</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./normalization.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Normalization and variance stabilization</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendices.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Appendices</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#understanding-the-input" id="toc-understanding-the-input" class="nav-link active" data-scroll-target="#understanding-the-input"> <span class="header-section-number">3.1</span> Understanding the input</a></li>
  <li><a href="#obtaining-fastq-files" id="toc-obtaining-fastq-files" class="nav-link" data-scroll-target="#obtaining-fastq-files"> <span class="header-section-number">3.2</span> Obtaining FASTQ files</a></li>
  <li><a href="#quantification" id="toc-quantification" class="nav-link" data-scroll-target="#quantification"> <span class="header-section-number">3.3</span> Quantification</a>
  <ul class="collapse">
  <li><a href="#x-chromium---cellranger" id="toc-x-chromium---cellranger" class="nav-link" data-scroll-target="#x-chromium---cellranger"> <span class="header-section-number">3.3.1</span> 10X Chromium - CellRanger</a></li>
  <li><a href="#starsolo" id="toc-starsolo" class="nav-link" data-scroll-target="#starsolo"> <span class="header-section-number">3.3.2</span> STARsolo</a></li>
  </ul></li>
  <li><a href="#kallisto-bustools" id="toc-kallisto-bustools" class="nav-link" data-scroll-target="#kallisto-bustools"> <span class="header-section-number">3.4</span> kallisto | bustools</a></li>
  <li><a href="#importing-into-r" id="toc-importing-into-r" class="nav-link" data-scroll-target="#importing-into-r"> <span class="header-section-number">3.5</span> Importing into R</a>
  <ul class="collapse">
  <li><a href="#importing-in-seurat" id="toc-importing-in-seurat" class="nav-link" data-scroll-target="#importing-in-seurat"> <span class="header-section-number">3.5.1</span> Importing in Seurat</a></li>
  <li><a href="#importing-in-singlecellexperiment" id="toc-importing-in-singlecellexperiment" class="nav-link" data-scroll-target="#importing-in-singlecellexperiment"> <span class="header-section-number">3.5.2</span> Importing in SingleCellExperiment</a></li>
  </ul></li>
  <li><a href="#adding-metadata" id="toc-adding-metadata" class="nav-link" data-scroll-target="#adding-metadata"> <span class="header-section-number">3.6</span> Adding metadata</a></li>
  <li><a href="#what-does-the-data-look-like" id="toc-what-does-the-data-look-like" class="nav-link" data-scroll-target="#what-does-the-data-look-like"> <span class="header-section-number">3.7</span> What does the data look like?</a></li>
  <li><a href="#final-note-on-objects" id="toc-final-note-on-objects" class="nav-link" data-scroll-target="#final-note-on-objects"> <span class="header-section-number">3.8</span> Final note on objects</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data input</span></h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<div class="callout-note callout callout-style-simple no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Relevant software:
</div>
</div>
<div class="callout-body-container callout-body">
<p><code>CellRanger</code> (CLI Linux)</p>
<p><code>STAR</code> (CLI Unix/Windows)</p>
<p><code>kb-python</code> (Python)</p>
<p><code>Seurat</code> (R CRAN)</p>
<p><code>DropletUtils</code> (R Bioconductor)</p>
<p><code>SingleCellExperiment</code> (R Bioconductor)</p>
</div>
</div>
<section id="understanding-the-input" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="understanding-the-input"><span class="header-section-number">3.1</span> Understanding the input</h2>
<p>Single-cell RNA-seq raw data come (mostly) in two flavours:</p>
<ul>
<li><p>3’ UMI (Unique Molecular Identifiers) - and more rarely 5’ UMI</p></li>
<li><p>Full length (can be UMI or not)</p></li>
</ul>
<p>The most important difference is in the throughput. UMI 3’ counts are compatible with droplet-based technologies, such as 10X Chromium, DropSeq and others. Full-length is compatible with plate-based (sorting) technologies such as SmartSeq, or microfluidic technologies such as Fluidigm C1.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>It is entirely possible to use UMI library preparation with a plate-based platform; UMIs were introduced to automatically de-duplicate reads and afford high precision. However, full length sequencing allows to capture more facets of gene expression such as alternative splicing or allele-specific expression.</p>
</div>
</div>
<p>Plate-based platforms are limited by the number of plates; each plate can hold at most up to 384 wells, meaning 384 single cells. Fluidigm chips can hold each up to 800 single cells.</p>
<p>Droplet-based technologies such as 10X Genomics Single Cell profiling (which combines droplets with microfluidic chips), up to 80,000 cells (320,000 with multiplexing in the High Throughput version) can be assayed in parallel.</p>
<p>Many projects generate 10X data, which uses a UMI 3’ library generation procedure.</p>
<p>The downside of using droplet-based platforms is that there is a trade-off between throughput (number of cells) and depth (number of reads) per cell. SmartSeq/C1 datasets will have fewer cells but a higher number of reads quantified, and therefore a higher number of genes detected.</p>
<p>When looking at UMI-based counts, you will have <strong>2</strong> files:</p>
<ul>
<li><p><strong>Read 1 (R1)</strong> contains the <strong>UMI</strong> + <strong>sample barcode,</strong> i.e.&nbsp;a sequence that is <strong>unique to the droplet</strong> (ideally the cell) and <strong>unique to the transcript</strong></p></li>
<li><p><strong>Read 2 (R2)</strong> contains the <strong>transcript read</strong>, i.e.&nbsp;a sequence that is <strong>unique to the transcript.</strong></p></li>
</ul>
<p>The combination of both reads gives the quantification of expression of a transcript within a droplet. The phenomenon of “empty droplets” happens when, for a given barcode, all of its R1 contains the barcode, and R2 contains the index, with no transcript read. This means that the droplet was sequenced and amplified but it did not contain any RNA.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>As far as I can tell by looking at unfiltered data, an extremely high number of reads goes into sequencing empty droplets.</p>
</div>
</div>
</section>
<section id="obtaining-fastq-files" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="obtaining-fastq-files"><span class="header-section-number">3.2</span> Obtaining FASTQ files</h2>
<p>Let’s start by a common situation: you want to download some publicly available data to reanalyze it, and you want to start from the raw data.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Most of the deposited data has already been processed by some standard quantification pipeline, making this step redundant most of the time. However, there are good reasons to start from the raw data again, e.g.&nbsp;if you are comparing datasets that have been processed by different pipelines, or if you need to use different transcript annotations.</p>
</div>
</div>
<p>Ideally you would use the server, as it has access to a faster bandwidth and its the final destination of most of these files.</p>
<p>The server has the <code>sratoolkit</code> set of tools installed. This means that, if we want to download some external dataset that is deposited on the Sequence Read Archive (SRA), we only need to use <code>prefetch</code>.</p>
<p>First, we need to know what samples we need. Let’s say we want to download the raw data (FASTQ files) from this <a href="https://pubmed.ncbi.nlm.nih.gov/31316211/">publication</a> by Schirmer et al.&nbsp;We go to the “Related information” section, and visit the link to <a href="https://www.ncbi.nlm.nih.gov/sra/?linkname=pubmed_sra&amp;from_uid=31316211">SRA</a>. At this point we want to look for the <strong>RUN</strong> accession number, which starts with “SRR” and will be the argument of <code>prefetch</code>. Note the argument <code>-X 50G</code> to <code>prefetch</code> which raises to 50GB the maximum download size. This may be necessary for large files.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">prefetch</span> <span class="at">-X</span> 50GB  SRR9123032</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This dataset is comprised of 21 runs, so it is tedious to look for each of them manually. One solution is to go to the European Nucleotide Archive (ENA) <a href="https://www.ebi.ac.uk/ena/browser/">website</a>, look for the same run number in the search term box, and then select the whole project accession number (the link starting with “PRJNA”), leading to this <a href="https://www.ebi.ac.uk/ena/browser/view/PRJNA544731">page</a>.</p>
<p>From the project page we can select a report as a .TSV file, and use simple bash commands to isolate the column of interest removing the header:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 4 filereport_read_run_PRJNA544731_tsv.txt <span class="kw">|</span> <span class="fu">tail</span> <span class="at">-n</span> 21 <span class="op">&gt;</span> samples.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This file can be uploaded to the server in our working directory using <code>scp</code> or <code>rsync</code>.</p>
<p>Once we have the <code>samples.txt</code> file available, we run a loop using <code>screen</code> to download these files in parallel:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">less</span> samples.txt<span class="kw">`;</span> </span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span> <span class="ex">screen</span> <span class="at">-dmS</span> <span class="st">"</span><span class="va">$i_prefetch</span><span class="st">"</span> <span class="kw">`</span><span class="ex">prefetch</span> <span class="at">-X</span> 50G <span class="va">$i</span><span class="kw">`;</span> </span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This will start a download in the background, which will create a new directory in your <code>home</code> named <code>ncbi</code>. The structure of the directory is:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">ncbi/public/sra</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And the <code>.sra</code> files will be stored there.</p>
<p>Once the <code>.sra</code> files have been downloaded, we have to unpack them to get our FASTQ files. This is done by running <code>fastq-dump</code> on each <code>.sra</code> file, remembering to add the <code>--split-files</code> argument to separate paired end reads (or UMI-barcode reads) and the <code>--gzip</code> argument to compress the files.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/ncbi/public/sra/</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="pp">*</span>.sra<span class="kw">;</span> </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span> <span class="ex">screen</span> <span class="at">-dms</span> <span class="st">"</span><span class="va">$i_dump</span><span class="st">"</span> <span class="kw">`</span><span class="ex">fastq-dump</span> <span class="va">$i</span> <span class="at">--split-files</span> <span class="at">--gzip</span><span class="kw">`;</span> </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout-warning callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>Do not delete the <code>.sra</code> files until you are sure that they are not corrupted.</p>
</div>
</div>
</section>
<section id="quantification" class="level2" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="quantification"><span class="header-section-number">3.3</span> Quantification</h2>
<p>You have downloaded your FASTQ files, or your collaborator has sent them to you. Now you need to transform these raw data into transcript counts, so that you can import them into R and start with the analysis.</p>
<p>There are a few options available, depending on the platform you are working on, the type of raw data, and the resources at your disposal (in terms of memory and processing power).</p>
<section id="x-chromium---cellranger" class="level3" data-number="3.3.1">
<h3 data-number="3.3.1" class="anchored" data-anchor-id="x-chromium---cellranger"><span class="header-section-number">3.3.1</span> 10X Chromium - CellRanger</h3>
<p>Data created through a 10X Chromium platform will be available in a format that is already amenable to quantification through <code>CellRanger</code>. Unfortunately, <code>CellRanger</code> is only available for Linux. An installation of <code>CellRanger</code> is available on the server.</p>
<p>The file naming needs to follow a specific structure:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">FileName_SampleNumber_L001_R1_001.fastq.gz</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="ex">FileName_SampleNumber_L001_R2_001.fastq.gz</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>where <code>FileName</code> is, for instance, the name of the run (<strong>SRR9123032</strong>), and <code>SampleNumber</code> is an internal convention (in this case, since this is the first sample, we can call it <strong>S1</strong>). R1 and R2 should be assigned to the barcode and the read+UMI respectively, but this in general is already taken care of by the submitter.</p>
<p>Filenames after a run of <code>fastq-dump</code> will instead be:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">SRR9123032_1.fastq.gz</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="ex">SRR9123032_2.fastq.gz</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>So we need to rename them (using <code>mv</code>) to match <code>CellRanger</code> specifications.</p>
<p>Assuming we are using the standard references that come packaged with <code>CellRanger</code>, we can run <code>cellranger count</code> specifying few parameters:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">cellranger</span> count <span class="at">--id</span><span class="op">=</span>S1 <span class="dt">\</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>--transcriptome=/data/data/RefAnnot/cellRanger/refdata-cellranger-GRCh38-3.0.0 <span class="dt">\</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>--fastqs=/data/data/SingleCell/giuseppe/schirmer/fastq/SRR9123032 <span class="dt">\</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>--sample=SRR9123032 <span class="dt">\</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>--chemistry=SC3Pv2 <span class="dt">\</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>--localcores=10 <span class="dt">\</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>--localmem=100 <span class="dt">\</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>--expect-cells=4000 <span class="dt">\</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>--include-introns</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Commenting the arguments:</p>
<ul>
<li><code>--id</code> should match the SampleNumber portion of the file name we assigned earlier</li>
<li><code>--transcriptome</code> points to the reference (either distributed by <code>CellRanger</code> or custom)</li>
<li><code>--fastqs</code> points to where the FASTQ files are</li>
<li><code>--sample</code> indicates the sample name, matching the FileName we assigned earlier</li>
<li><code>--chemistry=SC3Pv2</code> this depends on the publication, it is most likely either <code>v2</code> or <code>v3</code>. However the default is <code>auto</code> which means <code>CellRanger</code> will detect it.</li>
<li><code>--localcores=10</code> limit this to a reasonable number of cores</li>
<li><code>--localmem=100</code> same but considering the amount of RAM you have at your disposal</li>
<li><code>--expect-cells=4000</code> this is slightly above the default parameter, and it is informed by the publication</li>
<li><code>--include-introns</code> this is <strong>very important</strong> when dealing with single nuclei data, as most of the reads will come from pre-mRNA.</li>
</ul>
<p>Depending on the server load, 3 to 4 parallel runs (using <code>screen</code>) can be launched on the server.</p>
<p><code>CellRanger</code> will by default create .BAM (Binary Alignment Map) files. These are huge files that may not be useful as an end product. You can either include the <code>--no-bam</code> argument, or remove the BAM files after the run is finished.</p>
<section id="output-of-cellranger" class="level4" data-number="3.3.1.1">
<h4 data-number="3.3.1.1" class="anchored" data-anchor-id="output-of-cellranger"><span class="header-section-number">3.3.1.1</span> Output of CellRanger</h4>
<p><code>CellRanger</code> outputs 3 important sets of files:</p>
<ul>
<li><p>an HTML report showing so-called “knee plots” (more on that later)</p></li>
<li><p>a folder of unfiltered quantifications</p></li>
<li><p>a folder of filtered quantifications</p></li>
</ul>
<p>Filtering works by applying an algorithm that detects empty droplets with ambient RNA and removes them. Normally, only the filtered quantifications are imported in R. However, if you import unfiltered quantifications, you need to apply empty droplet removal by yourself. I detail how in the <a href="./qc.html">QC</a> chapter.</p>
</section>
</section>
<section id="starsolo" class="level3" data-number="3.3.2">
<h3 data-number="3.3.2" class="anchored" data-anchor-id="starsolo"><span class="header-section-number">3.3.2</span> STARsolo</h3>
<p>This is an updated version of the popular alignment tool <code>STAR</code>, which has been tweaked to align UMI and barcode-based reads, but can be used to align full-length reads as well.</p>
<p>I redirect you to the <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md">STARsolo github</a> page for some indications on how to use it.</p>
</section>
</section>
<section id="kallisto-bustools" class="level2" data-number="3.4">
<h2 data-number="3.4" class="anchored" data-anchor-id="kallisto-bustools"><span class="header-section-number">3.4</span> kallisto | bustools</h2>
<p>This is a completely different workflow, which makes use of tools built by Lior Pachter’s group: the <code>kallisto</code> pseudoaligner and <code>bustools</code> format specification for single cell data.</p>
<p>In a nutshell, rather than downloading FASTQ files first and then using memory-intensive reference-based alignment, this pipeline “streams” reads directly from a repository through the quantification pipeline, saving space, memory, and time.</p>
<p><code>kallisto</code> and <code>bustools</code> are available as a single Python package which can be installed via <code>pip</code> as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>pip install kb<span class="op">-</span>python</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="co">## if using python &gt;= 3</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>pip3 install kb<span class="op">-</span>python</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Or, if you are working on the server and using your own <code>conda</code> environment:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-c</span> bioconda kb-python</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then, we can follow the instructions at this <a href="https://sina.bio/2019/07/09/fasterq-to-count-matrices-for-single-cell-rna-seq/">tutorial</a>, which is a little bit out of date. I provide an updated version that should hopefully work for everyone.</p>
<p>First we want to download the necessary references for the <code>kallisto</code> pseudoaligner. These are the <em>transcript indices</em> (necessary for pseudo-alignment of reads) and a table mapping ENSEMBL <em>transcript IDs to gene IDs</em> and symbols. Following the example in the tutorial, we download indices and transcript references for <em>Mus musculus</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Download the transcript indices </span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-L</span> https://github.com/BUStools/getting_started/releases/download/getting_started/Mus_musculus.GRCm38.cdna.all.idx.gz <span class="at">-o</span> Mus_musculus.GRCm38.cdna.all.idx.gz</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co">## Download transcript to genes</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-L</span> https://github.com/BUStools/getting_started/releases/download/getting_started/transcripts_to_genes.txt <span class="at">-o</span> transcripts_to_genes.txt</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="co">## Extract references</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> Mus_musculus.GRCm38.cdna.all.idx.gz</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a><span class="co">## Assign variables</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="va">REFPATH</span><span class="op">=</span><span class="va">$PWD</span>/Mus_musculus.GRCm38.cdna.all.idx</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="va">TXPATH</span><span class="op">=</span><span class="va">$PWD</span>/transcripts_to_genes.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then, we can stream two FASTQ files from the Short Read Archive (SRA) from <a href="https://www.cell.com/immunity/fulltext/S1074-7613(19)30073-1">Koren et al.&nbsp;2019</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> mouse_data</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> mouse_data</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co">## Make named pipes (outputs)</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="fu">mkfifo</span> R1.gz R2.gz </span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-Ls</span> https://github.com/bustools/getting_started/releases/download/getting_started/SRR8599150_S1_L001_R1_001.fastq.gz  <span class="op">&gt;</span> <span class="va">$PWD</span>/R1.gz <span class="kw">&amp;</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-Ls</span> https://github.com/bustools/getting_started/releases/download/getting_started/SRR8599150_S1_L001_R2_001.fastq.gz  <span class="op">&gt;</span> <span class="va">$PWD</span>/R2.gz <span class="kw">&amp;</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="ex">kb</span> count <span class="at">-i</span> <span class="va">$REFPATH</span> <span class="at">-x</span> 10xv2 <span class="at">-g</span> <span class="va">$TXPATH</span> <span class="at">-t</span> 4 <span class="at">-o</span> <span class="va">$PWD</span> <span class="at">--cellranger</span> <span class="va">$PWD</span>/R1.gz <span class="va">$PWD</span>/R2.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In the code above, when calling <code>kb-count</code> we specify a few parameters:</p>
<ul>
<li><p><code>-i</code> the path to the index, which we assigned to a variable <code>$REFPATH</code></p></li>
<li><p><code>-x</code> the technology, in our case <code>10xv2</code></p></li>
<li><p><code>-g</code> the path to the transcript specification, assigned to <code>$TXPATH</code></p></li>
<li><p><code>-t</code> the number of threads - change according to your machine</p></li>
<li><p><code>-o</code> the output directory, which is the present working directory <code>$PWD</code></p></li>
<li><p><code>--cellranger</code> an option to create a <code>CellRanger</code>-compatible output</p></li>
<li><p>two positional arguments specifying the two named pipes (which contain the FASTQ files) created at the beginning.</p></li>
</ul>
<div class="callout-warning callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>Unlike <code>CellRanger</code> and <code>STARsolo</code>, <code>kb count</code> does not filter empty droplets. If you use this workflow, removing empty droplets is a part of processing that should be taken care of in R.</p>
</div>
</div>
<p>This process is quite fast and has a light footprint, meaning that with a few loops (or jobs, if you are using a High Performance Computing Cluster) should take care of large amounts of data in little time. The real advantage of using this workflow is that you can process raw data on a laptop without the need to have access to high performance machines.</p>
</section>
<section id="importing-into-r" class="level2" data-number="3.5">
<h2 data-number="3.5" class="anchored" data-anchor-id="importing-into-r"><span class="header-section-number">3.5</span> Importing into R</h2>
<p>Now that the processing is done, we want to import the counts into R.</p>
<p>The two most popular packages for single cell transcriptomics analysis in R are <code>Seurat</code> and the <code>SingleCellExperiment</code> (Bioconductor) tools.</p>
<p>I discuss their differences in the <a href="./pipeinechoice.html">pipeline choice</a> chapter, but for now I will show how to import <code>CellRanger</code> outputs (which can be created by the <code>kb count</code> pipeline as well) as either <code>Seurat</code> or <code>SingleCellExperiment</code> objects. Note that other outputs (such as <code>STARsolo</code>) can be imported just as easily.</p>
<section id="importing-in-seurat" class="level3" data-number="3.5.1">
<h3 data-number="3.5.1" class="anchored" data-anchor-id="importing-in-seurat"><span class="header-section-number">3.5.1</span> Importing in Seurat</h3>
<p><code>Seurat</code> can be installed from <code>CRAN</code> as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"Seurat"</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then, we can directly read the data as a sparse matrix (a memory-efficient representation of matrices where most of the entries are 0’s). Assuming the <code>CellRanger</code> output is in the <code>./cellranger</code> path:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">Read10X</span>(<span class="at">data.dir =</span> <span class="st">"./cellranger"</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>seurat_object <span class="ot">=</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="importing-in-singlecellexperiment" class="level3" data-number="3.5.2">
<h3 data-number="3.5.2" class="anchored" data-anchor-id="importing-in-singlecellexperiment"><span class="header-section-number">3.5.2</span> Importing in SingleCellExperiment</h3>
<p>For this import we need the <code>DropletUtils</code> package from Bioconductor:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"DropletUtils"</span>)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DropletUtils)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then, the creation of a <code>SingleCellExperiment</code> object only requires one line:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">read10xCounts</span>(<span class="at">samples =</span> <span class="st">"./cellranger"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="adding-metadata" class="level2" data-number="3.6">
<h2 data-number="3.6" class="anchored" data-anchor-id="adding-metadata"><span class="header-section-number">3.6</span> Adding metadata</h2>
<p>If you have other information available regarding the samples - patient of origin, RNA Integrity Number, date of collection, assigned cell type, experimental condition, etc - you should include it in the <code>colData</code> slot of the <code>SingleCellExperiment</code> object, or the <code>metadata</code> slot of the <code>Seurat</code> object. Metadata like this is usually reported at the cell-level, meaning that there has to be a correspondence between the columns of the assay and the rows of the metadata. Both approaches will fail if you try to create a <code>metadata</code>/<code>colData</code> whose rows don’t match with the columns of the <code>counts</code> table.</p>
<p>Assuming this cell-level metadata is in a tab-separated file named “cell_data.txt”, for the <code>SingleCellExperiment</code> previously generated you can add it as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>coldata <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">"cell_data.txt"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(sce) <span class="ot">&lt;-</span> coldata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Functions such as <code>colData</code>, <code>rowData</code>, <code>assay</code>, <code>reducedDim</code> are both S4 <em>getters</em> and <em>setters</em>, meaning they can either return the content of a slot (e.g.&nbsp;<code>pca &lt;- reducedDim(sce, name = "pca")</code> or overwrite it (e.g.&nbsp;<code>reducedDim(sce, name = "pca2") &lt;- pca2</code>).</p>
<p>In <code>Seurat</code>, getters and setters are separated. Weirdly enough the metadata getter is <code>[[]]</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>seurat_object <span class="ot">=</span> <span class="fu">AddMetaData</span>(seurat_object, <span class="at">metadata =</span> coldata)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>seurat_metadata <span class="ot">=</span> seurat_object[[]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="what-does-the-data-look-like" class="level2" data-number="3.7">
<h2 data-number="3.7" class="anchored" data-anchor-id="what-does-the-data-look-like"><span class="header-section-number">3.7</span> What does the data look like?</h2>
<p>Now is a good time to try to understand what the count data from a single cell transcriptomics experiment looks like, and introduce two important concepts.</p>
<p>The count table is a matrix (or 2D array) that contains <strong>m</strong> genes (usually rows) and <strong>n</strong> cells (columns). Every cell of this matrix will contain a positive integer value - the count for a particular gene in a particular cell. These matrices will have two intrinsic important features:</p>
<ul>
<li><p><strong>sparsity</strong>: because of the high throughput of these measurements (both in terms of numbers of cells and numbers of genes), the low absolute amount of input to be amplified during the library preparation, and the stochasticity of transcription, most of the genes will not be reliably quantified. In other words, many counts will be 0.<br>
<br>
Obviously, a gene that is not expressed will not be quantified. However, a gene that is expressed at a low or even mid level, and that would normally be measured by a bulk RNA-seq experiment, will be bypassed in a single-cell RNA-seq experiment. This is historically referred to as the “<strong>dropout</strong>” problem, because genes that would be normally detectable “drop out” of the measurement due to technical (or biological) reasons.<br>
<br>
Sparse data presents many challenges: does an average gene expression value make sense in the presence of a majority of 0 counts? Are the counts zero-inflated, or do they follow distributions more traditionally associated with RNA-seq? How noisy does the data get in the presence of a majority of 0’s?<br>
<br>
Several tools have been developed to deal with sparse data, and the single cell analysis pipelines make use of them to overcome its limitations. However, an important maxim to keep in mind when dealing with this data is that <strong>the absence of evidence is not the evidence of absence</strong>. A gene may not be measured even if it is expressed, and its activity may need to be inferred in many cases.</p></li>
<li><p><strong>high dimensionality</strong>: this is again a consequence of the high throughput, especially in terms of number of genes (features) measured. To better explain this concept, imagine you analyzed 100 cells and for each cell only measured 2 genes. You can plot these cells in a 2D space, positioning cells according to the conbined expression of these 2 genes. Greatly simplifying, you could divide cells in 4 possible “areas”: gene 1 high and gene 2 low, gene 2 high and gene 1 high, both high, both low.<br>
<br>
Now imagine adding a third gene. The 2D space becomes a 3D space, and the number of combinations increases by a factor of 2, resulting in 8 combinations (HHH, LLL, HLL, HLH, LLH, LHL, LHH, HHL). As we add genes - remember we are measuring several thousands of them! - the number of possible combination a single cell can inhabit increases.<br>
<br>
Some issues start arising: in the Summary I briefly mentioned that we need to relate cells to each other in terms of their distance in space. If these genes determine a coordinate system, then it is mathematically possible to calculate the Euclidean distance between two cells <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span> using the Pythagorean theorem in <span class="math inline">\(n\)</span> dimensions (<span class="math inline">\(n\)</span> genes):<br>
<br>
<span class="math display">\[
d(p, q) = \sqrt{(p_{1} - q_{1})^{2} + (p_{2} - q_{2})^{2} + (p_{3} - q_{3})^{2} + ... + (p_{n} - q_{n})^{2}}
\]</span><br>
<br>
However, <strong>as <em>n</em> increases, all points (cells) tend to become equidistant</strong>. This is known as the <strong>curse of dimensionality</strong> and it can be understood more in depth in this <a href="./appendices.html#appendix-1-the-curse-of-dimensionality">Appendix</a>. For now, it is sufficient to know that dealing with distances in high dimensional data “as is” is unlikely to yield sensible results, so that we need to find ways to reduce the dimensions.</p></li>
</ul>
<p>We refer to the <strong>m x n</strong> matrix as the <em>count</em> matrix, or <em>raw count</em> matrix, as it is untransformed and comes directly from the output of pre-processing pipelines. Some operations need to be performed using <em>raw counts</em>, whereas others require <em>normalized</em> or otherwise transformed counts. It is very important to understand when we need to use one or another as the statistical properties of counts (positive integers) are very different from those of transformed values that can take fractional and/or non-positive values.</p>
</section>
<section id="final-note-on-objects" class="level2" data-number="3.8">
<h2 data-number="3.8" class="anchored" data-anchor-id="final-note-on-objects"><span class="header-section-number">3.8</span> Final note on objects</h2>
<p>Ok, we got our input and we sort of understand what it looks like. But why do we need to import it into a <code>Seurat</code> or <code>SingleCellExperiment</code> object in the first place if they are just counts and metadata?</p>
<p>These objects have slots that hold different types of metadata or additional assays. For instance, in <code>SingleCellExperiment</code> objects:</p>
<ul>
<li><p><code>assays</code>, i.e.&nbsp;tables containing feature-level quantification where rows are features (genes) and columns are cells. <code>assays</code> may contain the raw <code>counts</code>, log-normalized <code>logcounts</code>, batch-corrected <code>corrected</code> assays, and any other assay the user may want to add.</p></li>
<li><p><code>colData</code>: a table containing cell-level information, such as QC metrics (number of transcripts detected, sum of reads, % of mitochondrial reads, batch of origin, assigned cell type, etc)</p></li>
<li><p><code>rowData</code>: a table containing feature-level information, such as alternative symbols, statistics, etc</p></li>
<li><p><code>metadata</code>: a slot for additional data that do not follow either a gene- or cell-major order</p></li>
<li><p><code>reducedDim</code>: a slot for dimensionality reductions such as <code>pca</code>, <code>umap</code>, <code>tsne</code>, <code>corrected</code></p></li>
</ul>
<p>and other slots.</p>
<p>Having everything in one object is not only “neater” but also allows to perform some operations easily: subsetting a <code>SingleCellExperiment</code> object by row - like you would with a matrix or data.frame - will subset all <code>assays</code> and <code>rowData</code>, as you are effectively subsetting over features; similarly, subsetting by column will subset <code>assays</code>, <code>reducedDim</code> and <code>colData</code> over cells.</p>
<p>Calling a named element from the list - e.g.&nbsp;<code>sce$patient</code> will subset from the <code>colData</code> directly.</p>
<p>Moreover, almost all functions from the Bioconductor family are written with the <code>SingleCellExperiment</code> will dispatch specific methods to this class. In other words, functions will automatically recognize a <code>SingleCellExperiment</code> object and perform certain operations that they wouldn’t do when called upon a matrix or a list. So, for instance, <code>runPCA</code> can be run on a matrix, thus returning an object containing the rotations, embeddings, and standard deviation calculations, or it can be run on a <code>SingleCellExperiment</code> object automatically using specific slots and populating the relevant <code>reducedDim</code> slot.</p>
<p>Many functions in fact take the whole object as an input, and return the same object with additional data in the relevant slots.</p>
<p><code>Seurat</code> objects have a very similar grammar, they can be subset in a nearly identical way and have their own methods dispatch.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./pipelinechoice.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Choice of pipeline</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./qc.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Quality Control</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>