<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-0.9.649">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Single Cell Analysis with R - 5&nbsp; Normalization and variance stabilization</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./appendices.html" rel="next">
<link href="./qc.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Normalization and variance stabilization</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Single Cell Analysis with R</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Summary</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pipelinechoice.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Choice of pipeline</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./datainput.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data input</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./qc.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Quality Control</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./normalization.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Normalization and variance stabilization</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendices.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Appendices</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#depth-normalization" id="toc-depth-normalization" class="nav-link active" data-scroll-target="#depth-normalization"> <span class="header-section-number">5.1</span> Depth normalization</a></li>
  <li><a href="#variance-stabilization" id="toc-variance-stabilization" class="nav-link" data-scroll-target="#variance-stabilization"> <span class="header-section-number">5.2</span> Variance stabilization</a>
  <ul class="collapse">
  <li><a href="#variance-stabilization-procedures" id="toc-variance-stabilization-procedures" class="nav-link" data-scroll-target="#variance-stabilization-procedures"> <span class="header-section-number">5.2.1</span> Variance stabilization procedures</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Normalization and variance stabilization</span></h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<div class="callout-important callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Relevant software:
</div>
</div>
<div class="callout-body-container callout-body">
<p><code style="padding: 0px 1px; border-width: 1px; border-style: solid; border-color: rgb(230, 230, 230) !important; border-image: initial; background-color: rgb(243, 243, 243) !important; font-family: Monaco, monospace !important; font-size: 9pt !important;">scuttle</code> (R Bioconductor)</p>
<p><code style="padding: 0px 1px; border-width: 1px; border-style: solid; border-color: rgb(230, 230, 230) !important; border-image: initial; background-color: rgb(243, 243, 243) !important; font-family: Monaco, monospace !important; font-size: 9pt !important;">sctransform</code> (R CRAN)</p>
<p><code style="padding: 0px 1px; border-width: 1px; border-style: solid; border-color: rgb(230, 230, 230) !important; border-image: initial; background-color: rgb(243, 243, 243) !important; font-family: Monaco, monospace !important; font-size: 9pt !important;">MatrixGenerics</code> (R Bioconductor)</p>
</div>
</div>
<section id="depth-normalization" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="depth-normalization"><span class="header-section-number">5.1</span> Depth normalization</h2>
<p>In the QC chapter we have seen a large variability in terms of total transcript/UMI count per cell. This is caused by both biological aspects - some cells make more and/or longer transcripts than others - and by technical aspects - transcripts from some cells have been captured or sequenced with a higher efficiency on average. Put it simply, it means that two very similar cells (same cell type/state, same phase of the cell cycle, same everything) may still look different just because 3000 UMI/reads total were counted for one and 6000 for the other. We say that these two cells are sequenced at a <em>different depth</em>.</p>
<p>This means that the naive comparison of cells can be distorted by this difference in depth: if we were to order cells by the variability of their genes without taking these differences into account, we would order them according to their respective depth and not any other biologically relevant feature.</p>
<p><strong>Normalization</strong> is a process that transforms the counts in a way that accounts for these differences, effectively rescaling them so that cells become more comparable. The per-cell scaling constant that is applied to equalize depths is commonly called the <strong>size factor</strong>.</p>
<p>One of the easiest ways to normalize cells is to divide each cell’s counts by a size factor equal to the sum total of counts (depth), and then multiply by a constant (10K or 1M).</p>
<p>Taking a raw count matrix <span class="math inline">\(X\)</span> with <span class="math inline">\(m\)</span> genes and <span class="math inline">\(n\)</span> cells, and indexing by <span class="math inline">\(i = \{1, 2, 3, ... ,m\}\)</span> and <span class="math inline">\(j = \{1, 2, 3, …, n\}\)</span> , the depth-normalized matrix <span class="math inline">\(X'\)</span> is given by:</p>
<p><span class="math display">\[
X'_{i,j} = \frac{X_{i,j}}{\sum_{i=1}^{m}x_{i,j}}\cdot10^{6}
\]</span></p>
<p>This is the depth-normalization as implemented in <code>Seurat</code>.</p>
<p>Different versions of depth normalization are implemented by <code>scuttle</code> in the <code>SingleCellExperiment</code> toolset. There are two main options:</p>
<ul>
<li><p><code>computeLibraryFctors()</code>: naive depth normalization where library sizes are scaled so that across cells their mean is equal to 1 (it is also possible to use the median or geometric mean). This is achieved by simply dividing the library sizes by their mean, then dividing the cells by these size factors. The assumptions behind this method are that there is no imbalanced differential expression between cells, i.e.&nbsp;up-regulation of a set of genes in one cell is matched by down-regulation of another set of cells in any other cell, so that there is no “excess differential expression”. This is the assumption that is used in bulk RNA-seq data analysis, but it is less likely to hold in single-cell data, especially when dealing with non-homogeneous datasets with different cell types and compositions and because of the prevalence of zero counts. This can also be referred to as “proportional fitting” (see below).</p></li>
<li><p><code>computeSumFactors()</code>: normalization by pooling and deconvolution: size factors are estimated by creating a reference “pseudocell” (averaging across all cells), pooling samll groups of cells together aggregating their counts by sum, and dividing the pooled counts by the pseudocell library size. This procedure is repeated iteratively including different cells in each iteration so that, eventually, a linear system of equations can be constructed which allows to derive the per-cell size factor. The advantages over the more naive size factor estimation are twofold: 1) the presence of 0 counts is alleviated by aggregating across cells, and 2) the compositional effects are taken into account by pre-clustering cells and using only cells within a cluster as pseudocell reference. This procedure is computationally more intensive.</p></li>
</ul>
</section>
<section id="variance-stabilization" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="variance-stabilization"><span class="header-section-number">5.2</span> Variance stabilization</h2>
<p>As important as it is, depth normalization is not enough to understand the “biological structure” of the data, i.e.&nbsp;to remove the technical confounding effects.</p>
<p>Counts from an RNA-seq experiment normally follow a Negative Binomial or a Poisson distribution depending on the capture and sequencing technology. These distributions arise due to several factors, both intrinsically technical (capture and reverse transcription efficiency in the presence of competition from highly expressed transcripts) and biological (bursty transcription and RNA degradation).</p>
<p>A notable feature of these distribution is that they exhibit a positive <strong>mean-variance</strong> <strong>relationship</strong>: as the mean expression of a gene increases, so does its variance.</p>
<p>This relationship holds even if we apply depth normalization.</p>
<p>Let’s simulate some data to convince ourselves, borrowing from examples in the <code>scran</code> package.</p>
<p>If we simulate counts from a Poisson distribution, we will have - in the un-normalized case - a linear trend where mean = variance:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MatrixGenerics)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: matrixStats</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'MatrixGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:matrixStats':

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># simulate Poisson distributed counts for 200 cells, 2000 genes with no structure</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">100</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>ncells <span class="ot">&lt;-</span> <span class="dv">200</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>ngenes <span class="ot">&lt;-</span> <span class="dv">2000</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate gene means: 2 to the power of a uniformly random number (min 2, max 10)</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># they will be the lambda parameter for the Poisson distribution</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>gene_means <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">^</span><span class="fu">runif</span>(ngenes, <span class="dv">2</span>, <span class="dv">10</span>)</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate counts</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>sim <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rpois</span>(ngenes<span class="sc">*</span>ncells, <span class="at">lambda =</span> gene_means), <span class="at">ncol =</span> ncells)</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot mean and variance</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> <span class="fu">rowMeans</span>(sim), <span class="at">y =</span> <span class="fu">rowVars</span>(sim), <span class="at">xlab =</span> <span class="st">"mean"</span>, <span class="at">ylab =</span> <span class="st">"variance"</span>)</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="normalization_files/figure-html/unnamed-chunk-1-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>After correction, the trend is still linear, albeit no longer equal (these normalized counts are no longer Poisson distributed):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Seurat epth normalization</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>sim_norm <span class="ot">=</span> <span class="fu">t</span>(<span class="fu">t</span>(sim)<span class="sc">/</span><span class="fu">colSums</span>(sim))<span class="sc">*</span><span class="fl">1e6</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot mean and variance after normalization</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> <span class="fu">rowMeans</span>(sim_norm), <span class="at">y =</span> <span class="fu">rowVars</span>(sim_norm), <span class="at">xlab =</span> <span class="st">"mean"</span>, <span class="at">ylab =</span> <span class="st">"variance"</span>)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>trend <span class="ot">=</span> <span class="fu">lm</span>(<span class="fu">rowVars</span>(sim_norm) <span class="sc">~</span> <span class="fu">rowMeans</span>(sim_norm))</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(trend, <span class="at">col =</span> <span class="st">"red"</span>)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">col =</span> <span class="st">"blue"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="normalization_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># scuttle depth normalization</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>sim_norm <span class="ot">=</span> <span class="fu">t</span>(<span class="fu">t</span>(sim)<span class="sc">/</span>(<span class="fu">colSums</span>(sim)<span class="sc">/</span><span class="fu">mean</span>(<span class="fu">colSums</span>(sim))))</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot mean and variance after normalization</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> <span class="fu">rowMeans</span>(sim_norm), <span class="at">y =</span> <span class="fu">rowVars</span>(sim_norm), <span class="at">xlab =</span> <span class="st">"mean"</span>, <span class="at">ylab =</span> <span class="st">"variance"</span>)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="normalization_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The NB distribution can be characterized by two parameters: <span class="math inline">\(\mu\)</span> and <span class="math inline">\(\phi\)</span>, i.e.&nbsp;mean and dispersion (<code>mu</code> and <code>size</code> in the <code>rnbinom()</code> function).</p>
<p>The dispersion parameter models the fact that in a NB distribution, compared to a Poisson, the variance increases at a higher rate than the mean; in other words the data are “overdispersed”. This dispersion parameter in our simulation is gene-specific and accounts for the aforementioned intrinsic technical variability.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># simulate Negative Binomial distributed counts for 200 cells, 2000 genes with no structure</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">100</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>ncells <span class="ot">&lt;-</span> <span class="dv">200</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>ngenes <span class="ot">&lt;-</span> <span class="dv">2000</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate gene means: 2 to the power of a uniformly random number (min 2, max 10)</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="co"># they will be the mu parameter for the NB distribution</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>gene_means <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">^</span><span class="fu">runif</span>(ngenes, <span class="dv">2</span>, <span class="dv">10</span>)</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate dispersions (size parameter)</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>dispersions <span class="ot">&lt;-</span> <span class="dv">10</span><span class="sc">/</span>gene_means <span class="sc">+</span> <span class="fl">0.2</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate counts</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>sim_nb <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnbinom</span>(ngenes<span class="sc">*</span>ncells, <span class="at">mu =</span> gene_means, <span class="at">size =</span> <span class="dv">1</span><span class="sc">/</span>dispersions), <span class="at">ncol =</span> ncells)</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot mean and variance</span></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> <span class="fu">rowMeans</span>(sim_nb), <span class="at">y =</span> <span class="fu">rowVars</span>(sim_nb), <span class="at">xlab =</span> <span class="st">"mean"</span>, <span class="at">ylab =</span> <span class="st">"variance"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="normalization_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Upon depth normalization:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Seurat depth normalization</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>sim_nb_norm_seurat <span class="ot">=</span> <span class="fu">t</span>(<span class="fu">t</span>(sim_nb)<span class="sc">/</span><span class="fu">colSums</span>(sim_nb))<span class="sc">*</span><span class="fl">1e6</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot mean and variance after normalization</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> <span class="fu">rowMeans</span>(sim_nb_norm_seurat), <span class="at">y =</span> <span class="fu">rowVars</span>(sim_nb_norm_seurat), </span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"mean"</span>, <span class="at">ylab =</span> <span class="st">"variance"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="normalization_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># scuttle depth normalization</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>sim_nb_norm_scuttle <span class="ot">=</span> <span class="fu">t</span>(<span class="fu">t</span>(sim_nb)<span class="sc">/</span>(<span class="fu">colSums</span>(sim_nb)<span class="sc">/</span><span class="fu">mean</span>(<span class="fu">colSums</span>(sim_nb))))</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot mean and variance after normalization</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> <span class="fu">rowMeans</span>(sim_nb_norm_scuttle), <span class="at">y =</span> <span class="fu">rowVars</span>(sim_nb_norm_scuttle), </span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"mean"</span>, <span class="at">ylab =</span> <span class="st">"variance"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="normalization_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>As you can see, removing the depth bias does not take care of removing the trend between variance and mean.</p>
<p>Recall that we need to order genes by their variances because we are interested in using the genes that exhibit the largest (biological) variability, as a way to filter out noise - genes that don’t change along biological axes of variation - and to reduce computational complexity by reducing the initial dimensions e.g.&nbsp;from 30,000 to 2,000.</p>
<p>If we were to select genes only based on their variance, ignoring the mean-variance relationship means we will choose highly expressed genes; among these there will be several housekeeping genes whose expression is unlikely to reveal the biological variability that is useful to separate cells in clusters or order them along differentiation processes.</p>
<p><strong>Variance stabilization</strong> is a procedure that tries to uncouple the relationship between variance and mean, so that the selection of highly variable genes does not necessarily include only the most expressed ones. Variance stabilization <em>per se</em> does not account for different depth, so it is usually applied after depth normalization. However, the <code>sctransform</code> procedure operates directly on raw counts, as it includes different depths in its modelling step.</p>
<section id="variance-stabilization-procedures" class="level3" data-number="5.2.1">
<h3 data-number="5.2.1" class="anchored" data-anchor-id="variance-stabilization-procedures"><span class="header-section-number">5.2.1</span> Variance stabilization procedures</h3>
<p>There are several methods for variance stabilization, and we can briefly explore/simulate some of them in conjunction with depth normalization:</p>
<ul>
<li><p><strong>log1p</strong>: this is a simple log2 transformation, where a “pseudocount” of 1 is added to each count before applying the logarithm to account for 0’s.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>sim_nb_log1p <span class="ot">=</span> <span class="fu">log2</span>(sim_nb <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> <span class="fu">rowMeans</span>(sim_nb_log1p), <span class="at">y =</span> <span class="fu">rowVars</span>(sim_nb_log1p), </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"mean"</span>, <span class="at">ylab =</span> <span class="st">"variance"</span>,</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"No depth normalization"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="normalization_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>sim_nb_norm_scuttle_log1p <span class="ot">=</span> <span class="fu">log2</span>(sim_nb_norm_scuttle <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> <span class="fu">rowMeans</span>(sim_nb_norm_scuttle_log1p), <span class="at">y =</span> <span class="fu">rowVars</span>(sim_nb_norm_scuttle_log1p), </span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"mean"</span>, <span class="at">ylab =</span> <span class="st">"variance"</span>,</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Scuttle depth normalization"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="normalization_files/figure-html/unnamed-chunk-7-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>sim_nb_norm_seurat_log1p <span class="ot">=</span> <span class="fu">log2</span>(sim_nb_norm_seurat <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> <span class="fu">rowMeans</span>(sim_nb_norm_seurat_log1p), <span class="at">y =</span> <span class="fu">rowVars</span>(sim_nb_norm_seurat_log1p), </span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"mean"</span>, <span class="at">ylab =</span> <span class="st">"variance"</span>,</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Seurat depth normalization"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="normalization_files/figure-html/unnamed-chunk-7-3.png" class="img-fluid" width="672"></p>
</div>
</div></li>
<li><p><strong>square root</strong>: variance is stabilized by taking the square root of counts. This does not require adding a pseudocount.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Poisson sqrt variance stabilization</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>sim_sqrt <span class="ot">=</span> <span class="fu">sqrt</span>(sim)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> <span class="fu">rowMeans</span>(sim_sqrt), <span class="at">y =</span> <span class="fu">rowVars</span>(sim_sqrt), <span class="at">xlab =</span> <span class="st">"mean"</span>, <span class="at">ylab =</span> <span class="st">"variance"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="normalization_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># NB sqrt variance stabilization</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>sim_nb_sqrt <span class="ot">=</span> <span class="fu">sqrt</span>(sim_nb)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> <span class="fu">rowMeans</span>(sim_nb_sqrt), <span class="at">y =</span> <span class="fu">rowVars</span>(sim_nb_sqrt), <span class="at">xlab =</span> <span class="st">"mean"</span>, <span class="at">ylab =</span> <span class="st">"variance"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="normalization_files/figure-html/unnamed-chunk-8-2.png" class="img-fluid" width="672"></p>
</div>
</div></li>
</ul>
<p>The <code>sqrt</code> transformation is mostly successful with Poisson distributed counts, but not with NB distributed counts.</p>
<ul>
<li><strong>PFlog1pPF</strong>: this is a term introduced by Booeshaghi et al.&nbsp;(2022), which consists in one round of depth normalization, followed by log1p transformation, and an additional round of depth normalization (or “proportional fitting” as they call it).</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>sim_nb_norm_scuttle_log1p <span class="ot">=</span> <span class="fu">log2</span>(sim_nb_norm_scuttle <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>sf <span class="ot">=</span> <span class="fu">colSums</span>(sim_nb_norm_scuttle_log1p)<span class="sc">/</span><span class="fu">mean</span>(<span class="fu">colSums</span>(sim_nb_norm_scuttle_log1p))</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>sim_nb_norm_log1p_pf <span class="ot">=</span> <span class="fu">t</span>(<span class="fu">t</span>(sim_nb_norm_scuttle_log1p)<span class="sc">/</span>sf)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>                           </span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> <span class="fu">rowMeans</span>(sim_nb_norm_log1p_pf), <span class="at">y =</span> <span class="fu">rowVars</span>(sim_nb_norm_log1p_pf), </span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"mean"</span>, <span class="at">ylab =</span> <span class="st">"variance"</span>,</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"PFlog1pPF"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="normalization_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<ul>
<li><strong>sctransform</strong> and the function <code>vst()</code>. This is a more complex procedure that implements both normalization and variance stabilization, together with highly variable gene selection (see later). Briefly, a regularized negative binomial regression model is applied to each gene, which allows the introduction of covariates (such as batch labels, sex, age, etc). Then the Pearson residuals (residuals divided by the square root of the expected value) for each gene are taken as “corrected” (normalized and stabilized) gene expression values. The variance of Pearson residuals can be used to rank genes as “highly variable”.</li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cell attributes from input UMI matrix: log_umi</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Variance stabilizing transformation of count matrix of size 2000 by 200</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Model formula is y ~ log_umi</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Get Negative Binomial regression parameters per gene</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Using 2000 genes, 200 cells</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |====================================================                  |  75%
  |                                                                            
  |======================================================================| 100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Second step: Get residuals using fitted parameters for 2000 genes</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |====================================================                  |  75%
  |                                                                            
  |======================================================================| 100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating gene attributes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Wall clock passed: Time difference of 1.619893 secs</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">rowMeans</span>(sim_sct<span class="sc">$</span>y), <span class="fu">rowVars</span>(sim_sct<span class="sc">$</span>y))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="normalization_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./qc.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Quality Control</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./appendices.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Appendices</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>