<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-0.9.649">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Single Cell Analysis with R - 2&nbsp; Choice of pipeline</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./datainput.html" rel="next">
<link href="./summary.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Choice of pipeline</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Single Cell Analysis with R</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Summary</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pipelinechoice.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Choice of pipeline</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./datainput.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data input</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./qc.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Quality Control</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./normalization.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Normalization and variance stabilization</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendices.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Appendices</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#seurat-or-singlecellexperiment" id="toc-seurat-or-singlecellexperiment" class="nav-link active" data-scroll-target="#seurat-or-singlecellexperiment"> <span class="header-section-number">2.1</span> Seurat or SingleCellExperiment?</a>
  <ul class="collapse">
  <li><a href="#seurat" id="toc-seurat" class="nav-link" data-scroll-target="#seurat"> <span class="header-section-number">2.1.1</span> Seurat</a></li>
  <li><a href="#singlecellexperiment" id="toc-singlecellexperiment" class="nav-link" data-scroll-target="#singlecellexperiment"> <span class="header-section-number">2.1.2</span> SingleCellExperiment</a></li>
  <li><a href="#final-thoughts" id="toc-final-thoughts" class="nav-link" data-scroll-target="#final-thoughts"> <span class="header-section-number">2.1.3</span> Final thoughts</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Choice of pipeline</span></h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<section id="seurat-or-singlecellexperiment" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="seurat-or-singlecellexperiment"><span class="header-section-number">2.1</span> Seurat or SingleCellExperiment?</h2>
<p>Before we delve into the details of a pipeline, it can be useful to look at the differences between the two most common (but not the only) single cell analysis toolkits: <code>Seurat</code> and the <code>SingleCellExperiment</code> tools.</p>
<p>This is not a benchmarking study, but rather a short writeup that reflects my experience using these tools and some of the advantages and disadvantages that I have personally encountered.</p>
<section id="seurat" class="level3" data-number="2.1.1">
<h3 data-number="2.1.1" class="anchored" data-anchor-id="seurat"><span class="header-section-number">2.1.1</span> Seurat</h3>
<p><code>Seurat</code> exists as a stand-alone library, created and maintained by the group of Rahul Satija at the New York Genome Center. It contains a multitude of routines and functions to from the import of raw files to the final analysis of several types of datasets.</p>
<p>It is, by all accounts, a one-stop solution to perform single cell RNA, spatial transcriptomics and multi-modal -omics analyses (such as integration with feature barcoding or ATAC-seq).</p>
<p><code>Seurat</code> is rather dependency-heavy, with some system dependencies (software that must be installed outside of R) that can make some installs a bit complicated. It heavily relies on Python libraries for several applications, which are called through the <code>reticulate</code> library.</p>
<p>The <code>Seurat</code> developers write extremely easy to follow documentation, with tutorials that make single cell analysis accessible to anyone with minimal coding experience. A dataset could be analyzed quite quickly with few lines of code, if you trust what that code is doing. Efficiency is indeed one of the attractive features of Seurat, although in many cases it has been out-performed by some other packages.</p>
<p>The issue with being a one-stop solution is that many choices are made by the developers before they can be made by the users. The use of modularity clustering algorithms - popularized by <code>Seurat</code> itself - is basically the only choice; theoretical discussions around the choice of parameters in the documentation are not many. In other words, <code>Seurat</code> takes away some of the complexity by also reducing some of the freedom and blackboxing many of its functions.</p>
<p>For a more technical - and slightly biased, probably - critique of <code>Seurat</code> you can read this <a href="https://github.com/LTLA/scRNAseq/issues/15#issuecomment-650648478">answer</a> by Aaron Lun on GitHub.</p>
</section>
<section id="singlecellexperiment" class="level3" data-number="2.1.2">
<h3 data-number="2.1.2" class="anchored" data-anchor-id="singlecellexperiment"><span class="header-section-number">2.1.2</span> SingleCellExperiment</h3>
<p>With <code>SingleCellExperiment</code> toolkit I identify a family of inter-related packages on Bioconductor that were developed to analyze all sorts of single cell -omics data. There is no single library that can do everything, but it all revolves around the <code>SingleCellExperiment</code> class object, which is derived by the older <code>SummarizedExperiment</code> class.</p>
<p>Packages that operate on and around this class usually take care of single aspects of the analysis.</p>
<p>For instance, <code>scater</code> implements quality control and visualizations; <code>bluster</code> contains a collection of clustering algorithms; <code>scran</code> contains functions for normalization, variance modelling, and differential expression; <code>batchelor</code> has functions to perform batch effect correction/integration; <code>AUCell</code> assigns cell identities; <code>slingshot</code> and <code>TSCAN</code> calculate pseudotime trajectories; etc.</p>
<p>The idea behind the Bioconductor philosophy is that the user is tasked with building their own pipeline using a highly modular collection of tools optimized for interoperability. This means that any developer can add their own package to the family and, depending on how well it works (on its own but especially with the other packages), it can become part of a more or less standardized pipeline.</p>
<p>Many developers working on many packages in parallel means that there is a high diversity of approaches, but also some fragmentation. Some Bioconductor packages may be abandoned by their developers, or may be extremely inefficient in their use of memory and/or CPU time.</p>
<p>In general, Bioconductor packages are very well maintained and highly efficient and interoperable. Some contributions to Bioncoductor may be less efficient - for instance, not all packages have implemented parallelization or out-of-memory operations - but they still have the benefit of working with the same objects and within the same set of rules.</p>
<p>The documentation available for Bioconductor is, most of the time, excellent; the OSCA book (from which this book heavily borrows for inspiration) is the most complete single source of information for single cell analyses as it contains not only case studies but also discussions on theoretical and practical aspects of the analysis at an appropriate level of depth.</p>
<p>Sometimes, however, Bioconductor packages lag a little behind innovations in <code>Seurat</code>. Spatial analysis was more developed in <code>Seurat</code> than in any other Bioconductor package; similarly, integration with ATAC-seq and other modalities was implemented earlier in <code>Seurat</code> (although it could still be achieved by using the right combination of Bioconductor packages).</p>
</section>
<section id="final-thoughts" class="level3" data-number="2.1.3">
<h3 data-number="2.1.3" class="anchored" data-anchor-id="final-thoughts"><span class="header-section-number">2.1.3</span> Final thoughts</h3>
<p>Many analysis tools in R try to cater to both crowds, implementing functions that can automatically recognize either <code>Seurat</code> or <code>SingleCellExperiment</code> objects. Moreover, both approaches try to maintain some level of interoperability between them (e.g.&nbsp;functions in <code>Seurat</code> to convert to <code>SingleCellExperiment</code> objects) and, to a certain extent, similar functions.</p>
<p>To summarize, here is a small table highlighting the strengths of both approaches.</p>
<table class="table">
<colgroup>
<col style="width: 29%">
<col style="width: 70%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;"><code>Seurat</code></th>
<th style="text-align: center;"><code>SingleCellExperiment</code></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">User friendliness</td>
<td style="text-align: center;">Highly customizable pipelines</td>
</tr>
<tr class="even">
<td style="text-align: center;">Good documentation (tutorials)</td>
<td style="text-align: center;">Great documentation (OSCA)</td>
</tr>
<tr class="odd">
<td style="text-align: center;">Efficiency</td>
<td style="text-align: center;">Efficiency (some)</td>
</tr>
<tr class="even">
<td style="text-align: center;">One-stop solution</td>
<td style="text-align: center;">Modularity and interoperability</td>
</tr>
<tr class="odd">
<td style="text-align: center;">Wide array of formats (spatial, multi-modal)</td>
<td style="text-align: center;">Many applications for transcriptomics not covered by Seurat (pseudotime, velocity, GRN reconstruction, etc)</td>
</tr>
<tr class="even">
<td style="text-align: center;">Single developer - cohesive product</td>
<td style="text-align: center;">Diversity - many opportunities but sometimes patchy</td>
</tr>
</tbody>
</table>
<p>I personally prefer the <code>SingleCellExperiment</code> family for its modularity, clarity of code and diversity, but I often use <code>Seurat</code> for some functions and approaches that may in some cases be faster, more efficient or give a better result.</p>
<p>Moreover, for Python users out there there is a whole other universe revolving around <code>scanpy</code> and <code>scVI</code>, collectively termed <code>scVerse</code>.</p>
<p>Throughout this book I will mostly focus on <code>SingleCellExperiment</code> solutions and, where convenient or interesting, point out the <code>Seurat</code> equivalent.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./summary.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Summary</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./datainput.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data input</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>