<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-0.9.649">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Single Cell Analysis with R - 4&nbsp; Quality Control</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./normalization.html" rel="next">
<link href="./datainput.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Quality Control</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Single Cell Analysis with R</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Summary</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pipelinechoice.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Choice of pipeline</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./datainput.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data input</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./qc.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Quality Control</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./normalization.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Normalization and variance stabilization</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendices.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Appendices</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#things-to-watch-out-for" id="toc-things-to-watch-out-for" class="nav-link active" data-scroll-target="#things-to-watch-out-for"> <span class="header-section-number">4.1</span> Things to watch out for</a></li>
  <li><a href="#basic-initial-qc-plots" id="toc-basic-initial-qc-plots" class="nav-link" data-scroll-target="#basic-initial-qc-plots"> <span class="header-section-number">4.2</span> Basic initial QC plots</a>
  <ul class="collapse">
  <li><a href="#total-counts-and-detected-genes" id="toc-total-counts-and-detected-genes" class="nav-link" data-scroll-target="#total-counts-and-detected-genes"> <span class="header-section-number">4.2.1</span> Total counts and detected genes</a></li>
  <li><a href="#barcode-ranks---knee-plots" id="toc-barcode-ranks---knee-plots" class="nav-link" data-scroll-target="#barcode-ranks---knee-plots"> <span class="header-section-number">4.2.2</span> Barcode ranks - knee plots</a></li>
  </ul></li>
  <li><a href="#removing-empty-droplets-and-ambient-rna" id="toc-removing-empty-droplets-and-ambient-rna" class="nav-link" data-scroll-target="#removing-empty-droplets-and-ambient-rna"> <span class="header-section-number">4.3</span> Removing empty droplets and ambient RNA</a></li>
  <li><a href="#mitochondrial-content" id="toc-mitochondrial-content" class="nav-link" data-scroll-target="#mitochondrial-content"> <span class="header-section-number">4.4</span> Mitochondrial content</a></li>
  <li><a href="#distribution-based-thresholding" id="toc-distribution-based-thresholding" class="nav-link" data-scroll-target="#distribution-based-thresholding"> <span class="header-section-number">4.5</span> Distribution-based thresholding</a></li>
  <li><a href="#doublet-identification" id="toc-doublet-identification" class="nav-link" data-scroll-target="#doublet-identification"> <span class="header-section-number">4.6</span> Doublet identification</a></li>
  <li><a href="#batch-aware-qc" id="toc-batch-aware-qc" class="nav-link" data-scroll-target="#batch-aware-qc"> <span class="header-section-number">4.7</span> Batch-aware QC</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Quality Control</span></h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<div class="callout-note callout callout-style-simple no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Relevant software:
</div>
</div>
<div class="callout-body-container callout-body">
<p><code>DropletUtils</code> (R Bioconductor)</p>
<p><code>scater</code> (R Bioconductor)</p>
<p><code>scuttle</code> (R Bioconductor)</p>
<p><code>scran</code> (R Bioconductor)</p>
<p><code>scDblFinder</code> (R Bioconductor)</p>
<p><code>scRNAseq</code> (R Bioconductor)</p>
<p><code>SingleCellExperiment</code> (R Bioconductor)</p>
</div>
</div>
<p>Once an object has been created in R, you have to make sure that you will be working with good data.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Most of these QC issues come from a sensible assessment of data, but in many cases whether a particular value represents a poor quality cell varies according to the source of data, the type of technology and platform, the intended application, etc. These should be considered as general guidelines and not strict rules to abide by.</p>
</div>
</div>
<section id="things-to-watch-out-for" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="things-to-watch-out-for"><span class="header-section-number">4.1</span> Things to watch out for</h2>
<p>In the previous chapter we saw that a lot of sequencing reads will be assigned to empty droplets. Removing those is just the beginning. Here is an incomplete list of things that you should assess critically:</p>
<ul>
<li><strong>balanced experimental design</strong><br>
This is not something that you can fix post-sequencing, but you can help your collaborator with at the time of planning. Experiments should be well designed, meaning that the size, timing and grouping of experimental samples avoids the introduction of uninteresting variation. Controls should be processed along with cases, in the same round; they can be split in different rounds only if for every control there is a corresponding case. Processing of samples should not be so long that the time difference between the first and the last sample consists of several hours. Sex, age, strain/ethnicity/background, should all be balanced (unless they are the subject of the study).</li>
<li><strong>ambient RNA contamination (aka “the soup”).</strong><br>
Cells may have been lysed in the culture medium/dissociation solution before they were captured in droplets/wells, releasing the content of their cytoplasms. This now cell-free RNA can still be included in droplets/wells, captured, amplified, and sequenced. Ambient RNA contamination can be a big issue for differential expression studies.</li>
<li><strong>high relative mitochondrial gene expression<br>
</strong>A cell that is expressing a high level of mitochondrial RNA is a cell that was already undergoing apoptosis or other forms of cell death, as the mitochondrial transcripts would have been released in the cytoplasm. It is a common conception in the field that, by discarding cells with a high % of mitochondrial transcripts, you are actually discarding suffering/dying cells that would confuse the analysis. However, some cell types (such as myocytes) can be rich in mitochondria owing to their big energy requirements. When</li>
<li><strong>low/high detection of transcripts<br>
</strong>Cells with too few reads or too many reads are usually discarded as damaged/ambient and doublets, respectively. Small cells, however, may have fewer RNA molecules.</li>
<li><strong>effect of the cell cycle<br>
</strong>Cell populations that are actively cycling may present different transcriptional profiles according to the phase of the cell cycle that they are in. If this effect is not accounted for, you could be mislead to think that there are 3 different cell populations (one for every detected phase); in reality the population is one, but it is cycling in an unsynchronized fashion. Moreover, when analyzing single nucleus RNA-seq, the presence of transcripts that are exclusively cytoplasmic or mitochondrial may point to poor library preparation, as the cytoplasm has not been stripped from the fixed tissues well enough, possibly creating confounding effects.</li>
<li><strong>High <em>MALAT1</em> expression<br>
</strong>In humans <em>MALAT1</em> is transcribed into a long non-coding RNA that is systematically and reliably quantified by most poly-A capture-based sequencing methods, effectively soaking up many sequencing reads. As a highly expressed gene, it may also appear as highly variable, thus creating the illusion that it is a biologically relevant marker. Evidence suggests that <em>MALAT1</em> is another stress-associated marker rather than an actual biological one. However, there are some cases (e.g.&nbsp;some cancers and leukaemias) in which <em>MALAT1</em> may play a more important function.</li>
<li><strong>number of cells<br>
</strong>Very trivially but importantly, the effective number of cells is important to be able infer properties from a biological system. Samples which have too few cells will fail to recapitulate the cell type composition of a tissue, and will most likely be over-clustered or overfitted by most algorithms. Collections of samples with highly unequal numbers of cells require additional steps to be included to account for these differences.</li>
<li><strong>low/absent gene quantification<br>
</strong>Chances are most of the genes will not be detected at all in your dataset. This does not mean that they were not expressed, but only that they were not measured due to low input and competition for sequencing reagent from more abundant transcripts. The absence of evidence is not the evidence of absence, and I will discuss a little more the problem of dropouts later on. Additionally, datasets in which only few genes are quantified and with a lot of noise may not be reliable.</li>
<li><strong>doublets<br>
</strong>Any single cell isolation technique will present the risk of incorporating more than one single cell per library preparation, meaning the same barcode will be shared by two different cells. These are called doublets and, if not accounted for, may look like intermediate cell states or entirely new cell populations.</li>
<li><strong>batch effects</strong><br>
Datasets coming from different individuals, different captures and/or different technologies will inevitably present batch effects. These are systematic biases in gene expression that we are not interested in, but nevertheless affect the analysis and interpretation. I will discuss batch effects later on, but they should already be tackled at the QC stage, since they can skew the distribution of metrics and the decision over which cells to keep and which to discard.</li>
</ul>
<p>There are other issues that can be more specific to certain technologies or tissues. For instance, the dissociation protocol may deplete a cell type, or it may activate an expression signature in a cell type (e.g.&nbsp;immune responses) that may look like a phenotype, etc.</p>
</section>
<section id="basic-initial-qc-plots" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="basic-initial-qc-plots"><span class="header-section-number">4.2</span> Basic initial QC plots</h2>
<p>Let’s use our <code>sce</code> object created through the <code>kallisto|bustools</code> pipeline in <a href="./datainput.html">Section 3 (data input)</a>.</p>
<div class="cell">

</div>
<p>As you will remember, it is an unfiltered dataset coming from a single mouse sample.</p>
<section id="total-counts-and-detected-genes" class="level3" data-number="4.2.1">
<h3 data-number="4.2.1" class="anchored" data-anchor-id="total-counts-and-detected-genes"><span class="header-section-number">4.2.1</span> Total counts and detected genes</h3>
<p>We start by applying per-cell quality control using the <code>scuttle</code> package:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scuttle)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">=</span> <span class="fu">addPerCellQC</span>(sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can then visualize the metrics using violin plots:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(<span class="at">object =</span> sce, <span class="at">y =</span> <span class="st">"sum"</span>) <span class="sc">+</span> <span class="fu">scale_y_log10</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Transformation introduced infinite values in continuous y-axis
Transformation introduced infinite values in continuous y-axis</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 7908 rows containing non-finite values (stat_ydensity).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 7908 rows containing missing values (geom_point).</code></pre>
</div>
<div class="cell-output-display">
<p><img src="qc_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(<span class="at">object =</span> sce, <span class="at">y =</span> <span class="st">"detected"</span>) <span class="sc">+</span> <span class="fu">scale_y_log10</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Transformation introduced infinite values in continuous y-axis</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Transformation introduced infinite values in continuous y-axis</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 7908 rows containing non-finite values (stat_ydensity).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 7908 rows containing missing values (geom_point).</code></pre>
</div>
<div class="cell-output-display">
<p><img src="qc_files/figure-html/unnamed-chunk-3-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>As you can see both the <code>sum</code> (total UMI per barcode) and the <code>detected</code> (number of transcripts with at least one UMI count) columns that were calculated by <code>addPerCellQC()</code> show that a very large portion of barcodes has only one read, then two, then three, etc. Due to the log-transformation in the plot scale, cells with <code>sum</code> = 0 were not plotted. We can still have an idea of their distribution:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">table</span>(sce<span class="sc">$</span>sum))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    0     1     2     3     4     5 
 7908 38805 17737 11049  5937  2856 </code></pre>
</div>
</div>
<p>We also want to check that the two variables follow the expected relationship:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(<span class="at">object =</span> sce, <span class="at">y =</span> <span class="st">"sum"</span>, <span class="at">x =</span> <span class="st">"detected"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="qc_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Expectedly, as total counts increase more genes are detected.</p>
</section>
<section id="barcode-ranks---knee-plots" class="level3" data-number="4.2.2">
<h3 data-number="4.2.2" class="anchored" data-anchor-id="barcode-ranks---knee-plots"><span class="header-section-number">4.2.2</span> Barcode ranks - knee plots</h3>
<p>It is difficult to establish what constitutes an empty droplet vs a “full” one just based on those plots. Ideally we would know where to establish a cutoff below which we consider each droplet to be empty.</p>
<p>To understand better where this cutoff may lie, we can plot the total UMI for each barcode (y axis) vs their rank (x axis), both log-transformed. This will give us a “knee plot” which should allow us to understand whether there are two more or less discrete populations (empty and full) and where the threshold may lie.</p>
<p>We take the code from the <code>DropletUtils</code> vignette.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>br.out <span class="ot">&lt;-</span> <span class="fu">barcodeRanks</span>(sce)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Making a plot.</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(br.out<span class="sc">$</span>rank, br.out<span class="sc">$</span>total, <span class="at">log=</span><span class="st">"xy"</span>, <span class="at">xlab=</span><span class="st">"Rank"</span>, <span class="at">ylab=</span><span class="st">"Total"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in xy.coords(x, y, xlabel, ylabel, log): 7908 y values &lt;= 0 omitted from
logarithmic plot</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>o <span class="ot">&lt;-</span> <span class="fu">order</span>(br.out<span class="sc">$</span>rank)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(br.out<span class="sc">$</span>rank[o], br.out<span class="sc">$</span>fitted[o], <span class="at">col=</span><span class="st">"red"</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="fu">metadata</span>(br.out)<span class="sc">$</span>knee, <span class="at">col=</span><span class="st">"dodgerblue"</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="fu">metadata</span>(br.out)<span class="sc">$</span>inflection, <span class="at">col=</span><span class="st">"forestgreen"</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"bottomleft"</span>, <span class="at">lty=</span><span class="dv">2</span>, <span class="at">col=</span><span class="fu">c</span>(<span class="st">"dodgerblue"</span>, <span class="st">"forestgreen"</span>), </span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend=</span><span class="fu">c</span>(<span class="st">"knee"</span>, <span class="st">"inflection"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="qc_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">metadata</span>(br.out)<span class="sc">$</span>inflection</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 186</code></pre>
</div>
</div>
<p>We see that a good threshold may lie at about 186 total UMIs for this dataset.</p>
</section>
</section>
<section id="removing-empty-droplets-and-ambient-rna" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="removing-empty-droplets-and-ambient-rna"><span class="header-section-number">4.3</span> Removing empty droplets and ambient RNA</h2>
<p>Having established a threshold, we can use it to call the <code>emptyDrops()</code> function from <code>DropletUtils</code>.</p>
<p>By specifying the threshold we are instructing the algorithm that whatever lies below 186 UMIs can be considered “ambient RNA”. Then, the algorithm compares the distribution of counts in each cell to the ambient distribution, and tests how significantly different it is by permutation testing. More details on the algorithm and testing regime can be found by typing <code>?emptyDrops</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>empty_droplets <span class="ot">&lt;-</span> <span class="fu">emptyDrops</span>(sce, <span class="at">lower =</span> <span class="dv">186</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>keep_droplets <span class="ot">&lt;-</span> empty_droplets<span class="sc">$</span>FDR <span class="sc">&lt;=</span> <span class="fl">0.01</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 92846</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(keep_droplets, <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2802</code></pre>
</div>
</div>
<p>We end up with 2801 “good quality” cells according to the content alone. This is a huge reduction considering we started with 92846 barcodes, and it means a lot of sequencing was inevitably wasted.</p>
<p>We can add this information to the <code>colData</code> of our object and plot the sum of UMI counts divided by ambient and non-ambient:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>empty <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">ifelse</span>(empty_droplets<span class="sc">$</span>FDR <span class="sc">&lt;=</span> <span class="fl">0.01</span>, <span class="st">"ok"</span>, <span class="st">"empty"</span>))</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>empty[<span class="fu">is.na</span>(sce<span class="sc">$</span>empty)] <span class="ot">=</span> <span class="st">"empty"</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(<span class="at">object =</span> sce, <span class="at">y =</span> <span class="st">"sum"</span>, <span class="at">x =</span> <span class="st">"empty"</span>, <span class="at">colour_by =</span> <span class="st">"empty"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="qc_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We also want to check that the number of permutations does not limit the ability of the algorithm to identify empty droplets:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="at">Sig =</span> keep_droplets, <span class="at">Limited =</span> empty_droplets<span class="sc">$</span>Limited)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Limited
Sig     FALSE TRUE
  FALSE   969    0
  TRUE   2693  109</code></pre>
</div>
</div>
<p>Since there are no barcodes that have <code>Limited == TRUE</code> and <code>Sig == FALSE</code> we can safely assume the iterations were sufficient, and the result is reliable.</p>
<p>We can remove all the empty barcodes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">=</span> sce[, <span class="fu">which</span>(sce<span class="sc">$</span>empty <span class="sc">==</span> <span class="st">"ok"</span>), drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="mitochondrial-content" class="level2" data-number="4.4">
<h2 data-number="4.4" class="anchored" data-anchor-id="mitochondrial-content"><span class="header-section-number">4.4</span> Mitochondrial content</h2>
<p>To calculate mitochondrial content we isolate mitochondrial genes - conveniently labelled in their gene symbol by starting with “mt-” - and calculate the relative expression resulting in the percentage of mitochondrial content.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>mito_genes <span class="ot">=</span> <span class="fu">grep</span>(<span class="st">"^mt-"</span>, <span class="fu">rowData</span>(sce)<span class="sc">$</span>Symbol)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">=</span> <span class="fu">addPerCellQCMetrics</span>(sce, <span class="at">subsets =</span> <span class="fu">list</span>(<span class="st">"mito"</span> <span class="ot">=</span> mito_genes))</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(<span class="at">object =</span> sce, <span class="at">y =</span> <span class="st">"subsets_mito_percent"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="qc_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="distribution-based-thresholding" class="level2" data-number="4.5">
<h2 data-number="4.5" class="anchored" data-anchor-id="distribution-based-thresholding"><span class="header-section-number">4.5</span> Distribution-based thresholding</h2>
<p>We can apply arbitrary thresholds, e.g.&nbsp;only retain cells with at least 400 total UMI counts, or below 5% mitochondrial content, but a more sophisticated approach consists in determining thresholds from the distributions themselves.<br>
</p>
<p>We use the <code>isOutlier()</code> function from <code>scuttle</code> to determine which barcodes are at least 3 MADs (Median Absolute Deviations) away from the median:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>low_lib <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(<span class="fu">log10</span>(sce<span class="sc">$</span>sum), <span class="at">type =</span> <span class="st">"lower"</span>, <span class="at">nmad=</span><span class="dv">3</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>low_genes <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(<span class="fu">log10</span>(sce<span class="sc">$</span>detected), <span class="at">type =</span> <span class="st">"lower"</span>, <span class="at">nmad=</span><span class="dv">3</span>)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>high_mt <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(sce<span class="sc">$</span>subsets_mito_percent, <span class="at">type =</span> <span class="st">"higher"</span>, <span class="at">nmad =</span> <span class="dv">4</span>)</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(<span class="at">LowLib=</span><span class="fu">sum</span>(low_lib), </span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>           <span class="at">LowNgenes=</span><span class="fu">sum</span>(low_genes), </span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>           <span class="at">HighMT =</span> <span class="fu">sum</span>(high_mt))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  LowLib LowNgenes HighMT
1     25        16     53</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>discard <span class="ot">&lt;-</span> low_lib <span class="sc">|</span> low_genes <span class="sc">|</span> high_mt </span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sce<span class="sc">$</span>discard)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
FALSE  TRUE 
 2738    64 </code></pre>
</div>
</div>
<p>We don’t need to remove many cells at this stage. We highlight cells marked for elimination:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(<span class="at">object =</span> sce, <span class="at">y =</span> <span class="st">"sum"</span>, <span class="at">colour_by =</span> <span class="st">"discard"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="qc_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We discard the marked barcodes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[, <span class="sc">!</span>sce<span class="sc">$</span>discard]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="doublet-identification" class="level2" data-number="4.6">
<h2 data-number="4.6" class="anchored" data-anchor-id="doublet-identification"><span class="header-section-number">4.6</span> Doublet identification</h2>
<p>We use the <code>scDblFinder</code> package to identify doublets in the data.</p>
<p>This package simulates artificial doublets by over-clustering cells, then mixing them together across clusters by summing the reads, and using these artificial doublets to create a classifier (based on the <code>XGBoost</code> libraries) that is able to assign a doublet probability to every cell, and their most likely origin.</p>
<p>One of the parameters of <code>scDblFinder</code> is the expected number of doublets, which depends on the technology and the initial amount of cells profiled. You can refer to <a href="https://kb.10xgenomics.com/hc/en-us/articles/360001378811-What-is-the-maximum-number-of-cells-that-can-be-profiled-">this link</a> for 10X Genomics Chromium platforms and library preparation protocols.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p><code>scDblFinder</code> and other doublet finding methods work well to identify <strong>heterotypic</strong> doublets, i.e.&nbsp;doublets in which cells come from different populations/types. If two cells of the same or similar type are included in a doublet - termed <strong>homotypic</strong> - they will not be called. However, homotypic doublets are accounted for since even though they cannot be distinguished they will still be generated.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scDblFinder)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">scDblFinder</span>(sce, <span class="at">clusters =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Clustering cells...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>8 clusters</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Creating ~5000 artificial doublets...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Dimensional reduction</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Evaluating kNN...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Training model...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>iter=0, 158 cells excluded from training.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>iter=1, 129 cells excluded from training.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>iter=2, 119 cells excluded from training.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Threshold found:0.528</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>133 (4.9%) doublets called</code></pre>
</div>
</div>
<p>Plotting the results:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(<span class="at">object =</span> sce, <span class="at">y =</span> <span class="st">"sum"</span>, <span class="at">x =</span>  <span class="st">"scDblFinder.class"</span>, <span class="at">colour_by =</span> <span class="st">"scDblFinder.class"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="qc_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>As expected, doublets tend to have a higher total UMI content.</p>
<div class="callout-important callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>Unlike empty droplets, we do not remove doublets at this stage. We flag them and see where they end up in downstream applications such as dimensional reduction and clustering.</p>
</div>
</div>
</section>
<section id="batch-aware-qc" class="level2" data-number="4.7">
<h2 data-number="4.7" class="anchored" data-anchor-id="batch-aware-qc"><span class="header-section-number">4.7</span> Batch-aware QC</h2>
<p>All the operations outlined above work well for a dataset derived from a single “batch”, i.e.&nbsp;a single patient, or capture, animal, etc. However, when dealing with multiple samples, it is important to take systematic biases into account.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>What is a batch effect?</strong><br>
“Whatever difference you do not care for but shows up nevertheless” would be a curt but still relevant answer.<br>
To be more precise while still being brief, whenever we sequence samples coming from different sources - patients, animals, labs, platforms, days of the week, machines - we introduce a systematic bias, i.e.&nbsp;a “batch effect”, that needs to be corrected and/or accounted for. If we don’t, we will find differences that are due to unintersting variability (e.g.&nbsp;between patients, or between days of the week) which has no biological interpretation.</p>
</div>
</div>
<p>To illustrate these differences, let’s download a small multi-sample single-cell RNA-seq dataset from Muraro et al.&nbsp;2016, using the <code>scRNAseq</code> package. This dataset contains pancreas cells from 4 donors, sorted in 8 plates each, and sequenced using the CEL-Seq2 technology, a full-length lower throughput platform.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scRNAseq)</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>muraro <span class="ot">&lt;-</span> <span class="fu">MuraroPancreasData</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>snapshotDate(): 2022-04-26</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>see ?scRNAseq and browseVignettes('scRNAseq') for documentation</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>loading from cache</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>see ?scRNAseq and browseVignettes('scRNAseq') for documentation</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>loading from cache</code></pre>
</div>
</div>
<p>We can see how cells from each donor have been sequenced in 8 plates:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(muraro<span class="sc">$</span>donor, muraro<span class="sc">$</span>plate)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     
       1  2  3  4  5  6  7  8
  D28 96 96 96 96 96 96 96 96
  D29 96 96 96 96 96 96 96 96
  D30 96 96 96 96 96 96 96 96
  D31 96 96 96 96 96 96 96 96</code></pre>
</div>
</div>
<p>Is there any systematic bias between plates and/or donors in the distribution of counts?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>mito_genes <span class="ot">=</span> <span class="fu">grep</span>(<span class="st">"^MT-"</span>, <span class="fu">rowData</span>(muraro)<span class="sc">$</span>symbol)</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>muraro <span class="ot">=</span> <span class="fu">addPerCellQCMetrics</span>(muraro, <span class="at">subsets =</span> <span class="fu">list</span>(<span class="st">"mito"</span> <span class="ot">=</span> mito_genes))</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(<span class="at">object =</span> muraro, <span class="at">x =</span> <span class="st">"donor"</span>, <span class="at">y =</span> <span class="st">"sum"</span>, <span class="at">colour_by =</span> <span class="st">"donor"</span>) <span class="sc">+</span> </span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>() <span class="sc">+</span></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Total UMI content by donor"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="qc_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(<span class="at">object =</span> muraro, <span class="at">x =</span> <span class="st">"donor"</span>, <span class="at">y =</span> <span class="st">"subsets_mito_percent"</span>, <span class="at">colour_by =</span> <span class="st">"donor"</span>) <span class="sc">+</span> </span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Mitochondrial content by donor"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="qc_files/figure-html/unnamed-chunk-19-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(<span class="at">object =</span> muraro, <span class="at">x =</span> <span class="st">"plate"</span>, <span class="at">y =</span> <span class="st">"sum"</span>, <span class="at">colour_by =</span> <span class="st">"plate"</span>) <span class="sc">+</span> </span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>() <span class="sc">+</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Total UMI content by plate"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="qc_files/figure-html/unnamed-chunk-19-3.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(<span class="at">object =</span> muraro, <span class="at">x =</span> <span class="st">"plate"</span>, <span class="at">y =</span> <span class="st">"subsets_mito_percent"</span>, <span class="at">colour_by =</span> <span class="st">"plate"</span>) <span class="sc">+</span> </span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Mitochondrial content by plate"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="qc_files/figure-html/unnamed-chunk-19-4.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>First things first: this dataset has no mitochondrial transcript quantification, nor does it require removal of empty droplets. Moreover, CEL-seq2 has an extremely low expected doublet rate (&lt; 1%, see Ding et al.&nbsp;Nature Biotechnology 2020).</p>
<p>You can see that, while plates are mostly similar, donors have different distributions of total read count. This means that applying a cutoff on the whole dataset without dividing it by donor will both include and remove several cells that, in a single datasets, would have been excluded/included.</p>
<p>The <code>batch</code> argument in <code>isOutlier()</code> allows to take this effect into account by applying the distribution-based cut-offs separately in each batch. In this case the <code>batch</code> argument is a vector containing the donor information, i.e.&nbsp;<code>muraro$donor</code>. In other cases it may just be sufficient to specify the name (character) of the column in <code>colData</code> for the function to know where to retrieve the batch labels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>low_lib <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(<span class="fu">log10</span>(muraro<span class="sc">$</span>sum), <span class="at">type =</span> <span class="st">"lower"</span>, <span class="at">nmad=</span><span class="dv">3</span>, <span class="at">batch =</span> muraro<span class="sc">$</span>donor)</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>low_genes <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(<span class="fu">log10</span>(muraro<span class="sc">$</span>detected), <span class="at">type =</span> <span class="st">"lower"</span>, <span class="at">nmad=</span><span class="dv">3</span>, <span class="at">batch =</span> muraro<span class="sc">$</span>donor)</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(<span class="at">LowLib=</span><span class="fu">sum</span>(low_lib), </span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>           <span class="at">LowNgenes=</span><span class="fu">sum</span>(low_genes))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  LowLib LowNgenes
1    250       285</code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>muraro<span class="sc">$</span>discard <span class="ot">&lt;-</span> low_lib <span class="sc">|</span> low_genes</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(muraro<span class="sc">$</span>discard)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
FALSE  TRUE 
 2787   285 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(<span class="at">object =</span> muraro, <span class="at">x =</span> <span class="st">"donor"</span>, <span class="at">y =</span> <span class="st">"sum"</span>, <span class="at">colour_by =</span> <span class="st">"discard"</span>) <span class="sc">+</span> </span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>() <span class="sc">+</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Total UMI content by donor"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="qc_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Similarly, other functions (such as <code>scDblFinder</code>) require you to specify the batch to make correct inferences about doublets.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./datainput.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data input</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./normalization.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Normalization and variance stabilization</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>